### DISCUSSION TITLE
Using ngx-mqtt to connect to AWS IOT Core

### DISCUSSION THREAD

**Discussion by 3AGSystemsShervin** [sclausen/ngx-mqtt#193] [Q&A]
Hi Team,

Has anyone been able to use this Angular wrapper to connect AWS IOT Core using Angular 9? I'm trying to subscribte to AWS IOT Core topic using my Angular 9 front end framework but I have had no luck. I cannot find a working example anywhere. Saw one for Angular 10 but I could not get it to work. Not even sure if its possible. Any help or guidance would be greatly appreciated. 

Thanks

**Comment by sclausen** [comment]
Could you please post what you found for angular 10 and maybe what you tried too (as a repository please)?

**Reply by 3AGSystemsShervin** [reply]
Thank you for your response. This is the example that I tried to follow for Angular 10 : 

**https://dev.to/eelayoubi/decoupling-an-app-using-aws-iot-577a**

I have also found this link which is using Angular 9 :

**https://github.com/aws/aws-iot-device-sdk-js-v2/issues/77**

Below is my environment, code and error I get when I follow the link for the Angular 9 setup :

https://github.com/aws/aws-iot-device-sdk-js-v2/issues/77


I'm also using Angular 9 but on a windows 10 OS. I have added the latest code mentioned here to my component but I get an error as soon as I uncomment the first line of code ( const clientBootstrap = new io.ClientBootstrap()).

Also when I add the suggested code to my polyfills.ts it gives a error below so I tried the code without making any changes to my polyfills.ts and import statements are not giving any errors and it compiles successfully.
(polyfills.ts:20 Uncaught TypeError: Cannot read properties of undefined (reading 'Buffer').
at Module../src/polyfills.ts (polyfills.ts:20)
at webpack_require (bootstrap:84)
at Object.2 (polyfills.ts:22)
at webpack_require (bootstrap:84)
at checkDeferredModules (bootstrap:45)
at Array.webpackJsonpCallback [as push] (bootstrap:32)
at polyfills.js:1s)

One thing to mention is for my project setup its recommended that I use Yarn however I have added other libraries using NPM and I have not had any issues mixing the two for installing libraries on same project. Not sure if that matters but I figured I mention it here

Below are the details to my environment, code and error I get :

**Environment**

Angular 9, aws-iot-device-sdk-v2, Windows 10

Node: 12.17.0
NPM: 6.14.4

**Code**

import { io, iot, mqtt } from 'aws-iot-device-sdk-v2';
import { Injectable } from '@angular/core';

@Injectable({
providedIn: 'root'
})
export class AWSIoTService {

constructor() { }

ngOnInit(): void {

const clientBootstrap = new io.ClientBootstrap();

// const myClient = iot.AwsIotMqttConnectionConfigBuilder
//   .new_with_websockets({
//     region: 'ap-southeast-1',
//     credentials_provider: undefined,
//     proxy_options: undefined
//   });

// const config = myClient.with_client_id('wsssss')
//   .with_clean_session(true)
//   .with_credentials('ap-southeast-1', 'xxxx', 'xxxx', '')
//   .with_endpoint('xxxx-ats.iot.ap-southeast-1.amazonaws.com')
//   .build();

// const client = new mqtt.MqttClient(clientBootstrap);

// const connection = client.new_connection(config);

// connection.on('connect', (session_present) => {
//   console.log(session_present);
//   console.log('connected');
//   connection.subscribe('/test/me/senpai', mqtt.QoS.AtLeastOnce, (topic, payload) => {
//     const decoder = new TextDecoder('utf8');
//     const message = decoder.decode(payload);
//     console.log(`message received: topic=${topic} message=${message}`);
//   }).then(subscription => {
//     return connection.publish(subscription.topic, 'NOTICE ME', subscription.qos);
//   });
// });

// connection.connect();
}
}

**Error**

ERROR Error: Uncaught (in promise): ReferenceError: exports is not defined
ReferenceError: exports is not defined
at Module../node_modules/aws-iot-device-sdk-v2/dist/iotidentity/iotidentityclient.js (app-main-main-module.js:36927)
at webpack_require (runtime.js:85)
at Object../node_modules/aws-iot-device-sdk-v2/dist/index.js (app-main-main-module.js:36777)
at webpack_require (runtime.js:85)
at Module../src/app/main/logicalunitview/logicalunitview.component.ts (app-main-main-module.js:151715)
at webpack_require (runtime.js:85)
at Module../src/app/main/main-routing.module.ts (app-main-main-module.js:152268)
at webpack_require (runtime.js:85)
at Module../src/app/main/main.module.ts (app-main-main-module.js:152373)
at webpack_require (runtime.js:85)
at resolvePromise (polyfills.js:13260)
at resolvePromise (polyfills.js:13203)
at polyfills.js:13329
at ZoneDelegate.invokeTask (polyfills.js:12710)
at Object.onInvokeTask (vendor.js:109700)
at ZoneDelegate.invokeTask (polyfills.js:12710)
at Zone.runTask (polyfills.js:12460)
at drainMicroTaskQueue (polyfills.js:12926)

**Reply by 3AGSystemsShervin** [reply]
@sclausen 

Latest update on this issue. I still cant get the library to work with older Angular project. As you can see from my other posts I was able to get the library to work with newer Angular project. One thing to note is that I cannot just simply upgrade the Angular in older project as it has a lot of code already which might break. So I need to get this working with older project. I have seen working examples on Stack Overflow with even angular 7, so I know its possible. I have spend a lot of time but have had no luck. I hope you can guide me in the right direction. Below are the notable files from both Angular projects I have created and the latest error I get when I add the module to my main module 

**Newer working solution:**

**package.json**

{
  "name": "aws-test",
  "version": "0.0.0",
  "scripts": {
    "ng": "ng",
    "start": "ng serve",
    "build": "ng build",
    "test": "ng test",
    "lint": "ng lint",
    "e2e": "ng e2e"
  },
  "private": true,
  "dependencies": {
    "@angular/animations": "~11.0.1",
    "@angular/common": "~11.0.1",
    "@angular/compiler": "~11.0.1",
    "@angular/core": "~11.0.1",
    "@angular/forms": "~11.0.1",
    "@angular/platform-browser": "~11.0.1",
    "@angular/platform-browser-dynamic": "~11.0.1",
    "@angular/router": "~11.0.1",
    "ngx-mqtt": "^9.0.5",
    "rxjs": "~6.6.0",
    "tslib": "^2.0.0",
    "zone.js": "~0.10.2"
  },
  "devDependencies": {
    "@angular-devkit/build-angular": "~0.1100.2",
    "@angular/cli": "~11.0.2",
    "@angular/compiler-cli": "~11.0.1",
    "@types/jasmine": "~3.6.0",
    "@types/node": "^12.11.1",
    "codelyzer": "^6.0.0",
    "jasmine-core": "~3.6.0",
    "jasmine-spec-reporter": "~5.0.0",
    "karma": "~5.1.0",
    "karma-chrome-launcher": "~3.1.0",
    "karma-coverage": "~2.0.3",
    "karma-jasmine": "~4.0.0",
    "karma-jasmine-html-reporter": "^1.5.0",
    "protractor": "~7.0.0",
    "ts-node": "~8.3.0",
    "tslint": "~6.1.0",
    "typescript": "~4.0.2"
  }
}



**tsconfig.json**

/* To learn more about this file see: https://angular.io/config/tsconfig. */
{
  "compileOnSave": false,
  "compilerOptions": {
    "baseUrl": "./",
    "outDir": "./dist/out-tsc",
    "forceConsistentCasingInFileNames": true,
    "strict": true,
    "noImplicitReturns": true,
    "noFallthroughCasesInSwitch": true,
    "sourceMap": true,
    "declaration": false,
    "downlevelIteration": true,
    "experimentalDecorators": true,
    "moduleResolution": "node",
    "importHelpers": true,
    "target": "es2015",
    "module": "es2020",
    "lib": [
      "es2018",
      "dom"
    ]
  },
  "angularCompilerOptions": {
    "strictInjectionParameters": true,
    "strictInputAccessModifiers": true,
    "strictTemplates": true
  }
}


**app.module.ts**

import { BrowserModule } from '@angular/platform-browser';
import { NgModule } from '@angular/core';

import { AppRoutingModule } from './app-routing.module';
import { AppComponent } from './app.component';

import { MqttModule, IMqttServiceOptions } from 'ngx-mqtt';

export const MQTT_SERVICE_OPTIONS: IMqttServiceOptions = {
  connectOnCreate: false,
  hostname: 'xxxxxxxx.iot.us-west-2.amazonaws.com',
  port: 443,
  path: '/mqtt'
};

@NgModule({
  declarations: [
    AppComponent
  ],
  imports: [
    BrowserModule,
    AppRoutingModule,
    MqttModule.forRoot(MQTT_SERVICE_OPTIONS),
  ],
  providers: [],
  bootstrap: [AppComponent]
})
export class AppModule { }



**Older project that throws error:**

**package.json**

{
  "name": "abp-zero-template",
  "version": "9.3.0",
  "license": "MIT",
  "angular-cli": {},
  "scripts": {
    "ng": "ng",
    "publish": "gulp build && ng build --prod",
    "start": "gulp buildDev && ng serve --host 0.0.0.0 --port 4200",
    "hmr": "gulp buildDev && ng serve --host 0.0.0.0 --port 4200 --hmr",
    "test": "gulp buildDev && ng test",
    "pree2e": "webdriver-manager update --standalone false --gecko false",
    "e2e": "gulp buildDev && protractor",
    "create-dynamic-bundles": "gulp buildDev",
    "nswag": "cd nswag/ && refresh.bat"
  },
  "private": false,
  "dependencies": {
    "@angular/animations": "^10.0.2",
    "@angular/cdk": "^10.0.1",
    "@angular/common": "^10.0.2",
    "@angular/compiler": "^10.0.2",
    "@angular/core": "^10.0.2",
    "@angular/forms": "^10.0.2",
    "@angular/platform-browser": "^10.0.2",
    "@angular/platform-browser-dynamic": "^10.0.2",
    "@angular/platform-server": "^10.0.2",
    "@angular/router": "^10.0.2",
    "@fullcalendar/core": "^5.1.0",
    "@microsoft/signalr": "^3.1.3",
    "@shopify/draggable": "^1.0.0-beta.8",
    "@swimlane/ngx-charts": "^14.0.0",
    "abp-ng2-module": "^6.2.0",
    "abp-web-resources": "^5.3.0",
    "adal-angular": "^1.0.17",
    "angular-gridster2": "^10.0.1",
    "angular-ng-autocomplete": "^2.0.1",
    "angular-oauth2-oidc": "^10.0.2",
    "angular2-counto": "^1.2.5",
    "angular2-text-mask": "^9.0.0",
    "animate.css": "^4.1.0",
    "apexcharts": "^3.27.2",
    "chart.js": "^2.9.3",
    "cookieconsent": "^3.1.1",
    "core-js": "^3.6.4",
    "d3": "^6.2.0",
    "d3-tile": "^1.0.0",
    "daterangepicker": "^3.1.0",
    "famfamfam-flags": "^1.0.0",
    "js-cookie": "^2.2.1",
    "just-compare": "^1.3.0",
    "localforage": "^1.7.3",
    "lodash": "^4.17.13",
    "moment": "^2.24.0",
    "moment-timezone": "^0.5.28",
    "msal": "^1.2.2",
    "ng-apexcharts": "^1.5.12",
    "ng-select": "^1.0.2",
    "ng2-file-upload": "^1.4.0",
    "ngx-bootstrap": "^5.6.1",
    "ngx-captcha": "^8.0.1",
    "ngx-image-cropper": "^3.1.4",
    "ngx-mqtt": "^9.0.5",
    "ngx-perfect-scrollbar": "^9.0.0",
    "ngx-spinner": "^9.0.1",
    "object-path": "^0.11.4",
    "primeicons": "^4.0.0",
    "primeng": "^9.1.2",
    "push.js": "^1.0.12",
    "quill": "^1.3.7",
    "rtl-detect": "^1.0.2",
    "rxjs": "^6.5.4",
    "simple-line-icons": "^2.4.1",
    "sweetalert2": "^9.10.7",
    "text-mask-addons": "^3.8.0",
    "topojson": "^3.0.2",
    "tslib": "^2.0.0",
    "webdriver-manager": "^12.1.7",
    "zone.js": "~0.10.2"
  },
  "devDependencies": {
    "@angular-devkit/build-angular": "~0.1000.0",
    "@angular/cli": "^10.0.0",
    "@angular/compiler-cli": "^10.0.2",
    "@angularclass/hmr": "^2.1.3",
    "@angularclass/hmr-loader": "^3.0.4",
    "@types/d3": "^6.2.0",
    "@types/jasmine": "~3.5.10",
    "@types/jasminewd2": "^2.0.8",
    "@types/lodash": "^4.14.149",
    "@types/moment-timezone": "^0.5.13",
    "@types/node": "^14.0.14",
    "codelyzer": "^6.0.0",
    "globby": "^11.0.1",
    "gulp": "^4.0.1",
    "gulp-clean-css": "^4.3.0",
    "gulp-concat": "^2.6.1",
    "gulp-less": "^4.0.1",
    "gulp-postcss": "^8.0.0",
    "gulp-uglify": "^3.0.2",
    "jasmine-core": "~3.5.0",
    "jasmine-spec-reporter": "~5.0.0",
    "karma": "~5.1.0",
    "karma-chrome-launcher": "~3.1.0",
    "karma-cli": "~2.0.0",
    "karma-coverage-istanbul-reporter": "~3.0.2",
    "karma-jasmine": "~3.3.0",
    "karma-jasmine-html-reporter": "^1.5.0",
    "merge-stream": "^2.0.0",
    "nswag": "13.6.2",
    "postcss-url": "^9.0.0",
    "protractor": "~7.0.0",
    "ts-node": "~8.10.2",
    "tslint": "~6.1.0",
    "typescript": "3.9.6"
  }
}


**tsconfig.json**

{
  "compilerOptions": {
    "declaration": false,
    "emitDecoratorMetadata": true,
    "experimentalDecorators": true,
    "lib": ["es6", "dom"],
    "mapRoot": "./",
    "module": "esnext",
    "skipLibCheck": true,
    "moduleResolution": "node",
    "outDir": "../dist/out-tsc",
    "sourceMap": true,
    "target": "es5",
    "typeRoots": ["../node_modules/@types"],
    "baseUrl": ".",
    "paths": {
      "@app/*": ["./app/*"],
      "@account/*": ["./account/*"],
      "@shared/*": ["./shared/*"],
      "@node_modules/*": ["../node_modules/*"],
      "@angular/*": ["../node_modules/@angular/*"],
      "@metronic/*": ["./assets/metronic/*"]
    }
  },
  "angularCompilerOptions": {
    "fullTemplateTypeCheck": true,
    "strictTemplates": true,
    "strictInputTypes": false,
    "strictAttributeTypes": false,
    "strictOutputEventTypes": false,
    "strictDomEventTypes": false,
    "strictLiteralTypes": false,
    "strictInjectionParameters": false
  }
}


**main.module.ts**

import { CommonModule } from '@angular/common';
import { NgModule } from '@angular/core';
import { FormsModule } from '@angular/forms';
import { AppCommonModule } from '@app/shared/common/app-common.module';
import { UtilsModule } from '@shared/utils/utils.module';
import { CountoModule } from 'angular2-counto';
import { ModalModule } from 'ngx-bootstrap/modal';
import { TabsModule } from 'ngx-bootstrap/tabs';
import { TooltipModule } from 'ngx-bootstrap/tooltip';
import { BsDropdownModule } from 'ngx-bootstrap/dropdown';
import { PopoverModule } from 'ngx-bootstrap/popover';
import { DashboardComponent } from './dashboard/dashboard.component';
import { MainRoutingModule } from './main-routing.module';
import { ChartComponent, NgxChartsModule } from '@swimlane/ngx-charts';
import { BsDatepickerConfig, BsDaterangepickerConfig, BsLocaleService } from 'ngx-bootstrap/datepicker';
import { BsDatepickerModule } from 'ngx-bootstrap/datepicker';
import { NgxBootstrapDatePickerConfigService } from 'assets/ngx-bootstrap/ngx-bootstrap-datepicker-config.service';
import { RealTimeDashboardComponent } from './realtimedashboard/realtimedashboard.component';
import { LogicalUnitComponent } from './logicalunit/logicalunit.component';
import { CreateLogicalUnitModalComponent } from './logicalunit/create-logicalunit-modal.component';
import { AppBsModalModule } from '@shared/common/appBsModal/app-bs-modal.module';
import {SelectModule} from 'ng-select';
import { EditLogicalUnitModalComponent } from './logicalunit/edit-logicalunit-modal.component';
import { CreateAlarmModalComponent } from './alarm/create-alarm-modal.component';
import { PaginatorModule } from 'primeng/paginator';
import { TableModule } from 'primeng/table';
import { AlarmReportingComponent } from './alarmreporting/alarmreporting.component';
import { LogicalUnitViewComponent } from './logicalunitview/logicalunitview.component';
import { AlarmViewComponent } from './alarmview/alarmview.component';
import { TimepickerModule } from 'ngx-bootstrap/timepicker';
import { TreeModule } from 'primeng/tree';
import { AutoCompleteModule, ButtonModule, SliderModule } from 'primeng';
import { CollapseModule } from 'ngx-bootstrap/collapse';
import { ViewOeeModalComponent } from './logicalunit/view-oee-modal.component';
import { EditAlarmModalComponent } from './alarm/edit-alarm-modal.component';
import { AlarmComponent } from './alarm/alarm.component';
import { LogicalUnitReportingComponent } from './logicalunitreporting/logicalunitreporting.component';
import { ScriptReportingComponent } from './scriptreporting.component/scriptreporting.component';
import { ViewDashboardConfigModalComponent } from './dashboardconfig/view-dashboardconfig-modal.component';
import { CreateDashboardConfigModalComponent } from './dashboardconfig/create-dashboardconfig-modal.component';
import { EditDashboardConfigModalComponent } from './dashboardconfig/edit-dashboardconfig-modal.component';
import { NgApexchartsModule } from 'ng-apexcharts';

import { MqttModule, IMqttServiceOptions } from 'ngx-mqtt';

export const MQTT_SERVICE_OPTIONS: IMqttServiceOptions = {
  connectOnCreate: false,
  hostname: 'xxxxxxxx-ats.iot.us-west-2.amazonaws.com',
  port: 443,
  path: '/mqtt'
};




NgxBootstrapDatePickerConfigService.registerNgxBootstrapDatePickerLocales();

@NgModule({
    imports: [
        CommonModule,
        FormsModule,
        ModalModule,
        TabsModule,
        TooltipModule,
        AppCommonModule,
        UtilsModule,
        MainRoutingModule,
        CountoModule,
        NgxChartsModule,
        BsDatepickerModule.forRoot(),
        TimepickerModule.forRoot(),
        BsDropdownModule.forRoot(),
        PopoverModule.forRoot(),
        AppBsModalModule,
        SelectModule,
        TableModule,
        TreeModule,
        PaginatorModule,
        SliderModule,
        AutoCompleteModule,
        ButtonModule,
        CollapseModule.forRoot(),
        NgApexchartsModule,
        MqttModule.forRoot(MQTT_SERVICE_OPTIONS),

    ],
    declarations: [
        DashboardComponent,
        RealTimeDashboardComponent,
        LogicalUnitComponent,
        LogicalUnitViewComponent,
        CreateLogicalUnitModalComponent,
        EditLogicalUnitModalComponent,
        EditAlarmModalComponent,
        CreateAlarmModalComponent,
        LogicalUnitReportingComponent,
        AlarmComponent,
        AlarmViewComponent,
        AlarmReportingComponent,
        ViewOeeModalComponent,
        ScriptReportingComponent,
        ViewDashboardConfigModalComponent,
        CreateDashboardConfigModalComponent,
        EditDashboardConfigModalComponent
    ],
    providers: [
        { provide: BsDatepickerConfig, useFactory: NgxBootstrapDatePickerConfigService.getDatepickerConfig },
        { provide: BsDaterangepickerConfig, useFactory: NgxBootstrapDatePickerConfigService.getDaterangepickerConfig },
        { provide: BsLocaleService, useFactory: NgxBootstrapDatePickerConfigService.getDatepickerLocale }
    ]
})
export class MainModule { }


**Compile Error thrown after module is added to main.module.ts**

ERROR in ./node_modules/ngx-mqtt/__ivy_ngcc__/fesm2015/ngx-mqtt.js 568:12-29
"export 'MqttClient' (imported as 'ɵngcc1') was not found in 'mqtt-browser'
ERROR in ./node_modules/ngx-mqtt/__ivy_ngcc__/fesm2015/ngx-mqtt.js 544:10-20
"export 'MqttClient' was not found in 'mqtt-browser'
ERROR in ./node_modules/ngx-mqtt/__ivy_ngcc__/fesm2015/ngx-mqtt.js 213:22-30
"export 'connect' (imported as '_connect') was not found in 'mqtt-browser'

**Reply by 3AGSystemsShervin** [reply]
@jariair @sclausen 

I was able to fix this issue by updating target on my tsconfig.josn file

**update it from**
"target": "es5",
**to**
"target": "es2015"

fixed the issue

**Comment by jariair** [comment]
I have got the AWS IOT working in Angular 11 (did work also in prior versions) by generating in the backend signed url for the connection. For the backend I use Python and aws_signv4_mqtt library. It needs IAM credentials having policy that allows certain client names (having fixed prefix) to connect to IOT endpoint and be able to Subscribe and Receive from certain topics. Those credentials can be used to sign the url for AWS IOT core websocket endpoint
`signed_url = aws_signv4_mqtt.generate_signv4_mqtt(iot_host, iot_host.split(".")[2], access_key, secret_key)`

This url is returned to Frontend Angular and used when connection to MQTT using transformWsUrl
`    this.mqttService.connect({
      url: this.mqtt_signed_url,
      port: 443,
      clientId: this.clientId,
      protocol: "wss",
      keepalive: 300,
      connectOnCreate: true,
      transformWsUrl: (url, options, client)=>{
        this.logger.log("MQTT/connect", "MQTT Transform WS Url")
        return this.mqtt_signed_url
      }
    }); `
    
Technically same could be achieved in Angular only by using AWS sdk's I would assume. I just did find it easier to sign the url in the backend. I know this is not full code sample but hopefully would guide towards one possible solution.

**Reply by 3AGSystemsShervin** [reply]
Thank you for your response:

I will try your solution with using AWS .NET SDK to get signed URL. Wanted to make sure a few things. Are you using ngx-mgtt for your angular library for connecting to AWS using the signed URL?

**Reply by 3AGSystemsShervin** [reply]
@jariair

Would it be possible for your to share your Angular code service class and component? After installation running "npm install ngx-mqtt --save" I get error when I try to import the module to my main module which my component is under. I get same error when I try to inject the service in my component. Wanted to see if you got this error using Angular 9:

"export 'MqttClient' (imported as 'ɵngcc1') was not found in 'mqtt-browser'
ERROR in ./node_modules/ngx-mqtt/__ivy_ngcc__/fesm2015/ngx-mqtt.js 544:10-20
"export 'MqttClient' was not found in 'mqtt-browser'
ERROR in ./node_modules/ngx-mqtt/__ivy_ngcc__/fesm2015/ngx-mqtt.js 213:22-30
"export 'connect' (imported as '_connect') was not found in 'mqtt-browser'

**Reply by jariair** [reply]
Sure, in my MqttClass there is some other service dependencies but I assume you can figure those out
[mqtt.service.ts.txt](https://github.com/sclausen/ngx-mqtt/files/7455757/mqtt.service.ts.txt)
[app.module.ts.txt](https://github.com/sclausen/ngx-mqtt/files/7455790/app.module.ts.txt)

In components where I need to connect/subscribe to topics 
`  
import { MQTTService } from 'src/app/services/mqtt.service';

constructor(
    private mqtt: MQTTService,
  ) {
this.mqtt.connect(["/feed1", "/feed2")

this.mqtt.eventFeed.subscribe(
      msg => {
        if( msg !== null ) {
          // Handle msg
        }
      }
}`

**Reply by 3AGSystemsShervin** [reply]
@jariair  Thanks again for your quick reply. I decided to create a new Angular project and add your suggestion there so I can at least get it working with latest version before I start to figure out why I'm getting those module errors in my project. I have simply added the connect logic into my component, just to make sure I can connect before adding anything else. Just note that the URL generated from the python code I have added substituted xxxxx's for access key, secret_key and on the host so its not shared over internet. But I have tried it with the real keys for my AWS account. I made sure that host endpoint and keys work by publishing to a topic using the "AmazonIotDataClient" from .NET SDK and checking those messages are successfully send by subscribing to that topic on the AWS MQTT test client on the web browser console. 

So right now I have no clue why I'm getting below error:

**mqtt.js:4441 WebSocket connection to 'wss://xxxxxxxxxxxxx-ats.iot.us-west-2.amazonaws.com/mqtt?X-Amz-Algorithm=AWS4-HMAC-SHA256&X-Amz-Credential=xxxxxxxxxxxxxxxxx%2F20211101%2Fus-west-2%2Fiotdevicegateway%2Faws4_request&X-Amz-Date=20211101T205236Z&X-Amz-SignedHeaders=host&X-Amz-Signature=xxxxxxxxxxxxxxxx' failed:** 

The error does not tell me what is wrong. Just says failed. I figure maybe you can try this code on your end and see if you can get connected successfully.

Appreciate your help. 

Below are the notable files in my new project: 


[app.component.txt](https://github.com/sclausen/ngx-mqtt/files/7456422/app.component.txt)
[app.module.txt](https://github.com/sclausen/ngx-mqtt/files/7456423/app.module.txt)
[package.txt](https://github.com/sclausen/ngx-mqtt/files/7456424/package.txt)

**Reply by 3AGSystemsShervin** [reply]
@jariair 

I figured out the problem. When I was generating the URL with python I just had to remove the https:// from the iot_host
 (https://xxxxx-ats.iot.us-west-2.amazonaws.com). For the .NET SDK it was required to have it in front of URL but here it was not. Very lucky I just tried that or I would be stock for sometime. Now I need to try to get it to work for my Angular 9 application.

Once again thanks for your help

**Comment by jariair** [comment]
Yes I'm using ngx-mqtt Angular library, currently using version 7.0.9 so would need to update at some point.

**Reply by 3AGSystemsShervin** [reply]
OK great. Seems like I cannot find anything for the .NET SDK to generate singed URL for AWS MQTT. There is something for generating for S3 but nothing for MQTT seems like. Might have to create a workaround using Python API to fetch the URL.

Thanks again for your time

**Comment by 3AGSystemsShervin** [comment]
@sclausen
 After installation running "npm install ngx-mqtt --save" I get error when I try to import the module to my main module which my component is under. I get same error when I try to inject the service in my component. Wanted to see if you got this error using Angular 9:

"export 'MqttClient' (imported as 'ɵngcc1') was not found in 'mqtt-browser'
ERROR in ./node_modules/ngx-mqtt/ivy_ngcc/fesm2015/ngx-mqtt.js 544:10-20
"export 'MqttClient' was not found in 'mqtt-browser'
ERROR in ./node_modules/ngx-mqtt/ivy_ngcc/fesm2015/ngx-mqtt.js 213:22-30
"export 'connect' (imported as '_connect') was not found in 'mqtt-browser'
